### PR TITLE
Vault aws iam auth method

### LINKED ISSUE #36: Use AWS IAM Auth Method for Vault authentication
This is feature request to implement login to Vault using AWS IAM auth method

Currently, the plugin logs into Vault using a Vault approle auth. This presents us with a number of issues:

1. The Vault approle design by itself creates a chicken and egg situation where we need to authenticate to Vault to fetch the `role-id` and `secret-id`, which is then used by systems to authenticate to that secret store. Albeit doable, it's difficult to automate the configuration of the plugin
2. A rotation of an approle `role-id` and `secret-id` requires intervention to refresh on TeamCity server side
3. `role-id` and `secret-id` are configured in the plugin to enable TeamCity to authenticate to Vault, and those information are stored on the filesystem of TeamCity server
    a. our instance of TeamCity acts as our CICD tool, and some of its build configurations will require secrets that grants access to production
    b. given `role-id` and `secret-id` are stored on file, should someone gain access to TeamCity server, that person could retrieve those information and generate tokens and gain access to the secrets

Given the reasons mentioned above and given the current state of the plugin, would it be possible to implement the authentication to Vault using AWS IAM auth method? We are willing to work on a Pull Request should this feature make sense in the plugin.

### PR DESCRIPTION
**Pull request by semiria**
Fixes #36 

Allow AWS IAM auth method to authenticate to Vault

### CODE REVIEW

File: server/src/main/resources/buildServerResources/editProjectConnectionVault.jsp [hunk 1]
```diff
@@ -163,10 +163,19 @@
...
+  $j(document).ready(function(){
+      let checked = $j('#aws-iam-auth').is(':checked')
+      $j("#approle_auth_endpoint").toggle(!checked);
+      $j("#approle_role_id").toggle(!checked);
+      $j("#approle_secret_id").toggle(!checked);
+  })
+
   $j('#aws-iam-auth').click(function() {
       $j("#approle_auth_endpoint").toggle(!this.checked);
+      $j("#approle_auth_endpoint input").val("")
```
Comment (VladRassokhin) [inline server/src/main/resources/buildServerResources/editProjectConnectionVault.jsp]: Please don't, it empties unsaved user input if someone accidentally clicked on that checkbox.
That should be done on server side

File: build.gradle [hunk 2]
```diff
@@ -56,8 +56,8 @@ subprojects {
     apply plugin: 'java-test-fixtures'
     apply plugin: 'kotlin'
 
-    sourceCompatibility = "1.6"
-    targetCompatibility = "1.6"
+    sourceCompatibility = "1.8"
```
Comment (VladRassokhin) [inline build.gradle:59]: Is that really necessary?

File: server/src/main/kotlin/org/jetbrains/teamcity/vault/server/VaultConnector.kt [hunk 3]
```diff
@@ -105,7 +109,35 @@ class VaultConnector(dispatcher: EventDispatcher<BuildServerListener>, private v
...
+                return getTokenFromAwsIamAuth(settings, trustStoreProvider)
+            }
+            return getTokenFromAppRole(template, settings)
+        }
+
+        private fun getTokenFromAwsIamAuth(settings: VaultFeatureSettings, trustStoreProvider: SSLTrustStoreProvider): Pair<String, String> {
+            val options = AwsIamAuthenticationOptions.builder()
+                    .credentialsProvider(InstanceProfileCredentialsProvider.getInstance()).build()
+
+            val template = createRestTemplate(settings, trustStoreProvider)
```
Comment (VladRassokhin) [inline server/src/main/kotlin/org/jetbrains/teamcity/vault/server/VaultConnector.kt]: Probably original `template` from `getToken(...)` could be used, no need to create another one

File: server/src/main/kotlin/org/jetbrains/teamcity/vault/server/VaultConnector.kt [hunk 4]
```diff
@@ -105,7 +109,35 @@ class VaultConnector(dispatcher: EventDispatcher<BuildServerListener>, private v
...
+                val aws = AwsIamAuthentication(options, template)
+                token = aws.login()
+                template.withVaultToken(token.token)
+                val response = template.getForEntity("/auth/token/lookup-self", VaultResponse::class.java)
+                val data = response.body.data
+                val accessor = data["accessor"].toString()
+
+                return token.token to accessor
+            } catch (e: Exception) {
+                throw Exception(e)
```
Comment (VladRassokhin) [inline server/src/main/kotlin/org/jetbrains/teamcity/vault/server/VaultConnector.kt]: Hm?

File: server/src/main/resources/buildServerResources/editProjectConnectionVault.jsp [hunk 5]
```diff
@@ -153,4 +162,20 @@
...
+
+  $j(document).ready(function(){
+      let checked = $j('#aws-iam-auth').is(':checked')
+      $j("#approle_auth_endpoint").toggle(!checked);
+      $j("#approle_role_id").toggle(!checked);
+      $j("#approle_secret_id").toggle(!checked);
+  })
+
+  $j('#aws-iam-auth').click(function() {
+      $j("#approle_auth_endpoint").toggle(!this.checked);
```
Comment (VladRassokhin) [inline server/src/main/resources/buildServerResources/editProjectConnectionVault.jsp]: Could be simplified by using class `auth_approle` on all relevant `tr` tags and `$j(".auth_approle").toggle(!this.checked)`

File: server/src/main/kotlin/org/jetbrains/teamcity/vault/server/VaultProjectConnectionProvider.kt [hunk 6]
```diff
@@ -64,14 +64,18 @@ class VaultProjectConnectionProvider(private val descriptor: PluginDescriptor) :
...
-                if (it[VaultConstants.FeatureSettings.ENDPOINT].isNullOrBlank()) {
-                    errors.add(InvalidProperty(VaultConstants.FeatureSettings.ENDPOINT, "Should not be empty"))
-                }
-                if (it[VaultConstants.FeatureSettings.ROLE_ID].isNullOrBlank()) {
-                    errors.add(InvalidProperty(VaultConstants.FeatureSettings.ROLE_ID, "Should not be empty"))
-                }
-                if (it[VaultConstants.FeatureSettings.SECRET_ID].isNullOrBlank()) {
-                    errors.add(InvalidProperty(VaultConstants.FeatureSettings.SECRET_ID, "Should not be empty"))
+
+                val awsIamAuth = it[VaultConstants.FeatureSettings.AWS_IAM_AUTH] ?: "false"
```
Comment (VladRassokhin) [inline server/src/main/kotlin/org/jetbrains/teamcity/vault/server/VaultProjectConnectionProvider.kt]: Please change to `?.toBoolean() ?: false` as it's used elsewhere (grep `toBoolean` on project)

**Review by VladRassokhin — CHANGES_REQUESTED** [review]
Thanks for your changes, please take a look on comments and edit commits history using interactive rebase.

Also I'm wondering whether it would be better to have radio buttons: `AppRole Auth` and `AWS IAM Auth`

Although I could finish it if it's OK and you don't want to invest more time

File: build.gradle (hunk 2 above)
Comment (semiria) [reply build.gradle:59]: Yes for it to work on my development environment, feel free to put that back to 1.6.

**Comment by semiria** [comment]
Thank you for your feedback @VladRassokhin. I've addressed most of your comments and left feedback on what I didn't. I think the radio button versus checkbox is a good idea, I went ahead and modified the code to reflect that comment.

I've sent you an invite to contribute on this fork, we'd be ok if you want to take over from here.

**Comment by semiria** [comment]
Any feedback @VladRassokhin ? As mentioned you can take over from here or I can address any other comments you may have, I'm open to both options.
Thanks!

**Comment by VladRassokhin** [comment]
Thank you for contribution, I've rebased, refactored and manually merged into master. Please try [latest build](https://teamcity.jetbrains.com/guestAuth/repository/download/TeamCityPluginsByJetBrains_TeamCityHashiCorpVaultPlugin_Build/.lastSuccessful/teamcity-hashicorp-vault-plugin.zip).

**Comment by semiria** [comment]
Thank you so much @VladRassokhin , I'll let you know if something comes up

**Pull request closed without merging**